pkg/updates/file.go

```diff
@@ -0,0 +1,49 @@
+package updates
+
+import (
+	"fmt"
+
+	"golang.org/x/exp/slices"
+)
+
+type Channel struct {
+	Name     string            `json:"name"`
+	Metadata map[string]string `json:"metadata,omitempty"`
+	Edges    EdgeSet           `json:"edges,omitempty"`
+	Nodes    []State           `json:"nodes,omitempty"`
+}
+
+type State struct {
+	ID        string `json:"id"`
+	Tag       string `json:"tag,omitempty"`
+	Migration string `json:"migration,omitempty"`
+	Phase     string `json:"phase,omitempty"`
+	Digest    string `json:"digest,omitempty"`
+}
+
+// UpdateGraph holds a graph of required update edges
+type UpdateGraph struct {
+	Channels []Channel `json:"channels,omitempty"`
+}
+
+func (g *UpdateGraph) ChannelForDatastore(datastore string) (string, error) {
+	for _, c := range g.Channels {
+		if c.Metadata["datastore"] == datastore {
+			return c.Name, nil
+		}
+	}
+	return "", fmt.Errorf("no channel found for datastore %q", datastore)
+}
+
+func (g *UpdateGraph) SourceForChannel(channel string) (Source, error) {
+	for _, c := range g.Channels {
+		if c.Name == channel {
+			return NewMemorySource(c.Nodes, c.Edges)
+		}
+	}
+	return nil, fmt.Errorf("no channel found with name %q", channel)
+}
+
+func (g *UpdateGraph) Copy() UpdateGraph {
+	return UpdateGraph{Channels: slices.Clone(g.Channels)}
```

Comment:
Why the need to Copy?

Reply:
each sync loop gets a copy of the config so that it gets a consistent view during reconciliation - the config can be updated out-of-band (the operator watches the file and re-loads if it changes)